cli/src/main.rs

```diff
@@ -1,8 +1,89 @@
+use db::PgPool;
+use tokio;
+use clap::{Parser, Subcommand};
+use db::sqlx;
+use std::fs;
+use std::io::Write;
+use uuid::Uuid; // Ensure you have the `uuid` crate in your Cargo.toml
+
+#[derive(Parser)]
+#[command(author, version, about, long_about = None)]
+struct KnowledgeArgs {
+    #[command(subcommand)]
+    command: Command,
+}
+
+#[derive(Subcommand)]
+enum Command {
+    Login,
+    Signup,
+    DisplayUsers,
+}
+
 #[tokio::main]
 async fn main() -> color_eyre::Result<()> {
-    let _db_pool = db::setup_db_pool().await?;
+    let args = KnowledgeArgs::parse();
+    let db_pool = db::setup_db_pool().await?;
+
+    match args.command {
+        Command::Login => {
+            println!("Login not implemented yet")
+        },
+        Command::Signup => {
+            sign_up(&db_pool).await?;
+        },
+        Command::DisplayUsers => {
+            display_users(&db_pool).await?;
+        }
+    }
+
+    Ok(())
+}
+
+async fn persist_auth_session(user_id: Uuid) -> color_eyre::Result<()> {
+    if fs::metadata("auth.txt").is_ok() {
+        return Err(color_eyre::eyre::eyre!("auth.txt already exists"));
+    }
+
+    let mut file = fs::File::create("auth.txt")?;
+    writeln!(file, "{}", user_id)?;
 
-    println!("Hello, world!");
+    Ok(())
+}
+
+async fn add_user(pool: &PgPool, username: &str) -> color_eyre::Result<()> {
+    let result = sqlx::query!(
+        "INSERT INTO Users (user_name) VALUES ($1) RETURNING user_id",
+        username
+    )
+    .fetch_one(pool)
+    .await?;
+
+    persist_auth_session(result.user_id).await?;
+
+    println!("User {} added successfully with ID {}!", username, result.user_id);
+    Ok(())
+}
+
+async fn display_users(pool: &PgPool) -> color_eyre::Result<()> {
+    let users = db::get_users(pool).await?;
+    if users.is_empty() {
+        println!("Nothing to display");
+    } else {
+        for user in users {
+            println!("User ID: {}, Username: {}", user.user_id, user.user_name);
+        }
+    }
+
+    Ok(())
+}
 
+async fn sign_up(pool: &PgPool) -> color_eyre::Result<()> {
+    println!("Enter a username");
+    let mut username = String::new();
+    std::io::stdin().read_line(&mut username).expect("Failed to read line");
+    let username = username.trim().to_string();
+    add_user(pool, &username).await?;
```

Comment:
`add_user` here takes a `&str`, which is what `trim()` returns! So you can probably remove the `to_string()` here!

The difference is that a `&str` is a reference to an existing piece of memory, and by calling `to_string` you are making an Owned String type, which needs a new bit of memory

So be leaving off the `to_string` we can save a small memory allocation!

Now if the compiler is mad at you with the reference, and this was to make it happy then never mind 😆 
Rust gets the most complicated with references, and thats where the infamous borrow checker comes in and gives lots of people problems. Cloning or `to_string`ing can get out of those problems and the cost of slightly worse theoritical performance. But OFTEN its not really worth it to fight the compiler, and a little `clone` or `to_string` just makes this easier lol

All that to say, if you can remove the `to_string` and it compiles lets do it! BUT if the compiler yells at ya, I'm 100% fine leaving this in lol

Reply:
My compiler is yelling at me 😆

Reply:
I tried removing it but the complier just wasn't having it